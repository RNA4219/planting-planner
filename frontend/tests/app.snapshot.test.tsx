import '@testing-library/jest-dom/vitest'
import { cleanup, screen, waitFor } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import type { MockInstance } from 'vitest'

type UseRecommendationsModule = typeof import('../src/hooks/useRecommendations')

import {
  fetchCrops,
  fetchRecommend,
  fetchRecommendations,
  renderApp,
  resetAppSpies,
} from './utils/renderApp'

describe('App snapshot', () => {
  let useRecommendationsModule: UseRecommendationsModule
  let useRecommendationsSpy: MockInstance

  beforeEach(async () => {
    resetAppSpies()
    fetchRecommend.mockRejectedValue(new Error('legacy endpoint disabled'))
    useRecommendationsModule = await import('../src/hooks/useRecommendations')
    useRecommendationsSpy = vi.spyOn(useRecommendationsModule, 'useRecommendations')
    fetchCrops.mockResolvedValue([
      {
        id: 1,
        name: 'トマト',
        category: 'flower',
      },
      {
        id: 2,
        name: 'レタス',
        category: 'leaf',
      },
    ])
    fetchRecommendations.mockResolvedValue({
      week: '2024-W30',
      region: 'temperate',
      items: [
        {
          crop: 'トマト',
          sowing_week: '2024-W28',
          harvest_week: '2024-W35',
          source: 'テストデータ',
          growth_days: 70,
        },
        {
          crop: 'レタス',
          sowing_week: '2024-W29',
          harvest_week: '2024-W32',
          source: 'テストデータ',
          growth_days: 45,
        },
      ],
      isMarketFallback: false,
    })
  })

  afterEach(() => {
    useRecommendationsSpy.mockRestore()
    cleanup()
    resetAppSpies()
  })

  test('初期表示をスナップショット保存する', async () => {
    await renderApp()
    await waitFor(() => {
      expect(screen.getByRole('row', { name: /トマト/ })).toBeInTheDocument()
    })

    const marketSelect = screen.getByRole('combobox', { name: '市場' })
    expect(marketSelect).toHaveAttribute('data-theme', 'market-national')
    expect(marketSelect.className).toContain('bg-market-national')
    expect(marketSelect).toHaveStyle({ color: 'rgb(255, 255, 255)' })

    const container = document.body.firstElementChild
    expect(container).not.toBeNull()
    expect(container?.querySelector('[class*="app__"]')).toBeNull()
    expect(container).toMatchSnapshot()
    expect(useRecommendationsSpy).toHaveBeenCalled()
  })
})
