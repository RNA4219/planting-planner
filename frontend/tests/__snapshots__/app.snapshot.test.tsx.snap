// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App snapshot > 初期表示をスナップショット保存する 1`] = `
<div>
  <div
    class="app"
  >
    <header
      class="app__header"
    >
      <h1
        class="app__title"
      >
        Planting Planner
      </h1>
      <form
        class="app__controls"
        novalidate=""
      >
        <label
          class="flex flex-col gap-2 text-sm font-medium text-slate-700"
        >
          <span>
            地域
          </span>
          <select
            aria-label="地域"
            class="rounded-md border border-slate-300 bg-white px-3 py-2 text-sm shadow-sm transition focus:border-green-500 focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2 disabled:cursor-not-allowed disabled:bg-slate-100"
            name="region"
          >
            <option
              value="cold"
            >
              寒冷地
            </option>
            <option
              value="temperate"
            >
              温暖地
            </option>
            <option
              value="warm"
            >
              暖地
            </option>
          </select>
        </label>
        <select
          aria-label="市場"
          name="marketScope"
        >
          <option
            value="national"
          >
            全国平均
          </option>
          <option
            value="city:tokyo"
          >
            東京都中央卸売
          </option>
          <option
            value="city:osaka"
          >
            大阪市中央卸売
          </option>
          <option
            value="city:nagoya"
          >
            名古屋市中央卸売
          </option>
        </select>
        <div
          class="app__controls-group"
        >
          <input
            aria-label="作物検索"
            id="search-input"
            name="search"
            placeholder="作物名・カテゴリで検索"
            type="search"
            value=""
          />
          <label
            class="app__week"
            for="week-input"
          >
            週
            <input
              id="week-input"
              inputmode="text"
              name="week"
              pattern="\\d{4}-W\\d{2}"
              placeholder="2024-W30"
              type="text"
              value="2024-W30"
            />
          </label>
          <div
            class="app__controls-actions"
          >
            <button
              type="submit"
            >
              この条件で見る
            </button>
            <button
              class="app__refresh"
              type="button"
            >
              更新
            </button>
          </div>
        </div>
      </form>
    </header>
    <main
      class="app__main"
    >
      <section
        aria-label="温暖地向けの推奨一覧（基準週: 2024-W30）"
        class="recommend space-y-4"
      >
        <div
          class="recommend__header flex flex-wrap items-center justify-between gap-3"
        >
          <div
            class="recommend__meta flex flex-wrap items-center gap-6 text-sm text-slate-600"
          >
            <span>
              対象地域: 
              温暖地
            </span>
            <span>
              基準週: 
              2024-W30
            </span>
          </div>
          <div
            aria-label="カテゴリ"
            class="inline-flex items-center gap-1 rounded-full bg-market-50 p-1"
            role="tablist"
          >
            <button
              aria-selected="true"
              class="rounded-full bg-transparent px-3 py-2 text-sm font-semibold text-market-700 transition-colors duration-200 hover:bg-market-100 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-market-400 aria-selected:bg-market-600 aria-selected:text-white"
              role="tab"
              tabindex="0"
              type="button"
            >
              葉菜
            </button>
            <button
              aria-selected="false"
              class="rounded-full bg-transparent px-3 py-2 text-sm font-semibold text-market-700 transition-colors duration-200 hover:bg-market-100 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-market-400 aria-selected:bg-market-600 aria-selected:text-white"
              role="tab"
              tabindex="-1"
              type="button"
            >
              根菜
            </button>
            <button
              aria-selected="false"
              class="rounded-full bg-transparent px-3 py-2 text-sm font-semibold text-market-700 transition-colors duration-200 hover:bg-market-100 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-market-400 aria-selected:bg-market-600 aria-selected:text-white"
              role="tab"
              tabindex="-1"
              type="button"
            >
              花き
            </button>
          </div>
        </div>
        <table
          aria-label="温暖地向けの推奨一覧（基準週: 2024-W30）"
          class="recommend__table w-full border-separate border-spacing-4"
        >
          <thead
            class="sr-only"
          >
            <tr>
              <th
                scope="col"
              >
                作物
              </th>
              <th
                scope="col"
              >
                期間
              </th>
              <th
                scope="col"
              >
                情報源
              </th>
            </tr>
          </thead>
          <tbody
            class="grid gap-4 sm:grid-cols-2 xl:grid-cols-3"
          >
            <tr
              aria-selected="false"
              class="recommend__row flex h-full flex-col gap-4 rounded-2xl border p-4 shadow-sm transition focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-market-accent focus-visible:ring-offset-2 bg-market-neutral/10 border-market-neutral/40 ring-0"
              data-testid="recommendation-card"
              tabindex="0"
            >
              <td
                class="flex items-start justify-between gap-3 align-top"
              >
                <div
                  class="flex items-start gap-2"
                >
                  <button
                    aria-label="トマトをお気に入りに追加"
                    aria-pressed="false"
                    class="inline-flex items-center justify-center text-xl leading-none transition-colors duration-200 bg-transparent border-0 cursor-pointer p-0 hover:text-amber-400 focus-visible:text-amber-400 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-amber-300 text-slate-300"
                    type="button"
                  >
                    ☆
                  </button>
                  <span
                    class="text-sm font-semibold text-slate-700"
                  >
                    トマト
                  </span>
                </div>
                <span
                  class="text-xs text-slate-500"
                >
                  カテゴリ: 
                  果菜類
                </span>
              </td>
              <td
                class="grid grid-cols-2 gap-2 text-xs text-slate-600 align-top"
              >
                <div>
                  <p
                    class="font-medium text-slate-500"
                  >
                    播種週
                  </p>
                  <p
                    class="text-sm text-slate-700"
                  >
                    2024-W28
                  </p>
                </div>
                <div>
                  <p
                    class="font-medium text-slate-500"
                  >
                    収穫週
                  </p>
                  <p
                    class="text-sm text-slate-700"
                  >
                    2024-W35
                  </p>
                </div>
              </td>
              <td
                class="text-xs text-slate-600 align-top"
              >
                <p
                  class="font-medium text-slate-500"
                >
                  情報源
                </p>
                <p
                  class="text-sm text-slate-700"
                >
                  テストデータ
                </p>
              </td>
            </tr>
            <tr
              aria-selected="false"
              class="recommend__row flex h-full flex-col gap-4 rounded-2xl border p-4 shadow-sm transition focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-market-accent focus-visible:ring-offset-2 bg-market-neutral/10 border-market-neutral/40 ring-0"
              data-testid="recommendation-card"
              tabindex="0"
            >
              <td
                class="flex items-start justify-between gap-3 align-top"
              >
                <div
                  class="flex items-start gap-2"
                >
                  <button
                    aria-label="レタスをお気に入りに追加"
                    aria-pressed="false"
                    class="inline-flex items-center justify-center text-xl leading-none transition-colors duration-200 bg-transparent border-0 cursor-pointer p-0 hover:text-amber-400 focus-visible:text-amber-400 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-amber-300 text-slate-300"
                    type="button"
                  >
                    ☆
                  </button>
                  <span
                    class="text-sm font-semibold text-slate-700"
                  >
                    レタス
                  </span>
                </div>
                <span
                  class="text-xs text-slate-500"
                >
                  カテゴリ: 
                  葉菜類
                </span>
              </td>
              <td
                class="grid grid-cols-2 gap-2 text-xs text-slate-600 align-top"
              >
                <div>
                  <p
                    class="font-medium text-slate-500"
                  >
                    播種週
                  </p>
                  <p
                    class="text-sm text-slate-700"
                  >
                    2024-W29
                  </p>
                </div>
                <div>
                  <p
                    class="font-medium text-slate-500"
                  >
                    収穫週
                  </p>
                  <p
                    class="text-sm text-slate-700"
                  >
                    2024-W32
                  </p>
                </div>
              </td>
              <td
                class="text-xs text-slate-600 align-top"
              >
                <p
                  class="font-medium text-slate-500"
                >
                  情報源
                </p>
                <p
                  class="text-sm text-slate-700"
                >
                  テストデータ
                </p>
              </td>
            </tr>
          </tbody>
        </table>
      </section>
      <section
        class="mt-8"
      >
        <h2>
          価格推移
        </h2>
        <p
          aria-live="polite"
          role="status"
        >
          作物を選択すると価格推移が表示されます。
        </p>
        <p
          class="mt-2 text-sm text-slate-500"
        >
          作物一覧で行をクリックすると、価格推移が表示されます。
        </p>
      </section>
    </main>
  </div>
</div>
`;
